File: app/src/main/java/yhb/dc/MainActivity.java
package yhb.dc;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import yhb.dc.common.CommonAdapter;
import yhb.dc.common.CommonViewHolder;
import yhb.dc.common.Demo;
import yhb.dc.common.LifeCycleActivity;

public class MainActivity extends LifeCycleActivity {

    private static final int debuggingDemoId = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        List<Class<? extends Activity>> activities = new ArrayList<>();

        try {
            PackageManager packageManager = getPackageManager();
            PackageInfo packageInfo = packageManager.getPackageInfo(
                    this.getPackageName(), PackageManager.GET_ACTIVITIES
            );
            Class autoJumpTo = null;
            for (ActivityInfo activityInfo : packageInfo.activities) {
                @SuppressWarnings("unchecked") Class<? extends Activity> activityClass
                        = (Class<? extends Activity>) Class.forName(activityInfo.name);
                Demo annotation = activityClass.getAnnotation(Demo.class);
                if (annotation == null) {
                    continue;
                }
                activities.add(activityClass);
                if (annotation.id() == debuggingDemoId) {
                    autoJumpTo = activityClass;
                }
            }

            if (autoJumpTo != null) {
                Intent intent = new Intent(this, autoJumpTo);
                startActivity(intent);
            }

        } catch (PackageManager.NameNotFoundException | ClassNotFoundException e) {
            e.printStackTrace();
        }

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new CommonAdapter<Class<? extends Activity>>(R.layout.item_activity, activities) {
            @Override
            public void convert(CommonViewHolder holder, final Class<? extends Activity> entity) {
                final Demo annotation = entity.getAnnotation(Demo.class);
                final String specifiedName = annotation == null ? null : annotation.name();
                holder.setText(R.id.text_view_activity_name, specifiedName == null || specifiedName.isEmpty()
                        ? entity.getSimpleName() : specifiedName);
                holder.bindOnClickListener(v -> {
                    Intent intent = new Intent(MainActivity.this, entity);
                    startActivity(intent);
                });
            }
        });
    }


}

File: app/src/main/java/yhb/dc/MainActivity.kt
package yhb.dc

import android.app.Activity
import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.view.Menu
import yhb.dc.common.CommonAdapter
import yhb.dc.common.CommonViewHolder
import yhb.dc.common.Demo
import yhb.dc.common.LifeCycleActivity

class MainActivity : LifeCycleActivity() {

    private val activities: MutableList<Class<Activity>> = ArrayList()
    private var recyclerView: RecyclerView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        activities.addAll(generateDemoList())
        recyclerView = findViewById(R.id.recycler_view)
        refreshList(activities)
    }

    private fun refreshList(demoList: List<Class<out Activity?>>) {
        val recyclerView = recyclerView ?: return
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = object : CommonAdapter<Class<out Activity>>(R.layout.item_activity, demoList) {
            override fun convert(holder: CommonViewHolder, clazz: Class<out Activity>?) {
                val entity = clazz ?: return
                holder.setText(R.id.text_view_activity_name, demoName(entity))
                holder.bindOnClickListener {
                    val intent = Intent(this@MainActivity, entity)
                    startActivity(intent)
                }
            }
        }
    }

    private fun demoName(demoClazz: Class<out Activity>): String {
        val annotation = demoClazz.getAnnotation(Demo::class.java) ?: null
        val specifiedName = annotation?.name
        return if (specifiedName.isNullOrEmpty()) demoClazz.simpleName else specifiedName
    }

    private fun generateDemoList(): List<Class<Activity>> {
        val activities: MutableList<Class<Activity>> = ArrayList()
        try {
            val packageManager = packageManager
            val packageInfo = packageManager.getPackageInfo(
                    this.packageName, PackageManager.GET_ACTIVITIES
            )
            var autoJumpTo: Class<*>? = null
            for (activityInfo in packageInfo.activities) {
                val activityClass = Class.forName(activityInfo.name) as Class<Activity>
                val annotation = activityClass.getAnnotation(Demo::class.java) ?: continue
                activities.add(activityClass)
                if (annotation.id == debuggingDemoId) {
                    autoJumpTo = activityClass
                }
            }
            if (autoJumpTo != null) {
                val intent = Intent(this, autoJumpTo)
                startActivity(intent)
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
        return activities
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.ab_search).actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String?): Boolean {
                refreshList(filterDemoList(newText))
                return true
            }

            override fun onQueryTextSubmit(query: String?): Boolean {
                return true
            }

            private fun filterDemoList(keyword: String?): List<Class<Activity>> {
                if (keyword.isNullOrEmpty()) {
                    return activities
                }
                return activities.filter { demoName(it).contains(keyword, ignoreCase = true) }.toCollection(ArrayList())
            }
        })
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(componentName))
        return true
    }

    companion object {
        private const val debuggingDemoId = -1
    }
}